adafruit_led_animation/animation.py

```diff
@@ -0,0 +1,585 @@
+
+import random
+from math import ceil, sin, radians
+from .color import BLACK, RAINBOW
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_LED_Animation.git"
+
+
+class Animation:
+    """
+    Base class for animations.
+    """
+    # pylint: disable=too-many-arguments
+    def __init__(self, pixel_object, speed, color, peers=None, paused=False):
+        self.pixel_object = pixel_object
+        self._speed_ns = 0
+        self.speed = speed  # sets _speed_ns
+        self._color = color
+        self._next_update = monotonic_ns()
+        self.pixel_object.auto_write = False
+        self.color = color
+        self.peers = peers if peers else []
+        self._paused = paused
+        self._time_left_at_pause = 0
+
+    def animate(self):
+        """
+        Call animate() from your code's main loop.  It will draw the animation draw() at intervals
+        configured by the speed property (set from init).
+
+        :return: True if the animation draw cycle was triggered, otherwise False.
+        """
+        if self._paused:
+            return False
+
+        now = monotonic_ns()
+        if now < self._next_update:
+            return False
+
+        self.draw()
+        if self.peers:
+            for peer in self.peers:
+                peer.draw()
+        self.show()
+        for peer in self.peers:
+            peer.show()
+
+        self._next_update = now + self._speed_ns
+        return True
+
+    def draw(self):
+        """
+        Animation subclasses must implement draw() to render the animation sequence.
+        """
+        raise NotImplementedError()
+
+    def show(self):
+        """
+        Displays the updated pixels.  Called during animates with changes.
+        """
+        self.pixel_object.show()
+
+    @property
+    def paused(self):
+        """
+        Whether the animation is paused.
+        """
+        return self._paused
+
+    @paused.setter
+    def paused(self, value):
+        if self._paused == value:
+            return
+
+        self._paused = value
+        if value:
+            self._time_left_at_pause = max(0, monotonic_ns() - self._next_update)
+        else:
+            self._next_update = monotonic_ns() + self._time_left_at_pause
+            self._time_left_at_pause = 0
+
+    def fill(self, color):
+        """
+        Fills the pixel object with a color.
+        """
+        self.pixel_object.fill(color)
+        self.show()
+
+    @property
+    def color(self):
+        """
+        The current color.
+        """
+        return self._color
+
+    @color.setter
+    def color(self, color):
+        if isinstance(color, int):
+            color = (color >> 16 & 0xff, color >> 8 & 0xff, color & 0xff)
+        self._color = color
+        self._recompute_color(color)
+
+    @property
+    def speed(self):
+        """
+        The animation speed in fractional seconds.
+        """
+        return self._speed_ns / 1000000000
```

Comment:
Since this is technically a "magic" number and appears more than once, consider pulling this into a module-level constant, like https://github.com/theacodes/Winterbloom-Sol-software/blob/ce2b8b07f19b5f5c20cd32fa50e2ef7e7007c603/winterbloom_sol/adsr.py#L30

Reply:
I had thought about that, but I am also concerned about performance.  I vaguely remember there's a constant thing in  CircuitPython that might be faster than a bare module variable, but I need to benchmark.  Having worked a lot in the CircuitPython core with a JLink and gdb lately makes me think about these things :)

I'll look into the performance of both.

Reply:
From my experience moving things to module-level variables doesn't impact performance, and might even help, considering that CP doesn't need to do an allocation for the module-level variable every time you use it.

If performance is a big concern we should do another pass at this once it's merged. There's lots of use of `for x in y`, `range`, and slicing that outright murder performance on CP compared to manipulating index variables.

Reply:
Another pass at it was already planned. The initial code was written to make it work. Making it better is the next step.

Reply:
@theacodes there is indeed plenty of room for performance fixes.  Two different things happened while writing this:  1) Making it work.  2) A lot of the code I wrote was written with _pixelbuf in mind, which perform faster with slices..  Running the code without _pixelbuf is, as you said, much slower.  Once we get this merged, there's a number of issues to file regarding improvements, additions, and design documentation.

Reply:
Removed the use of sin, however the code is still too slow for large strips as-is, but not as slow as it was - but that'll be easy to address in future PRs.

Reply:
I moved this to a constant in the module.